packages/lib-auth/src/client.js

```diff
@@ -0,0 +1,318 @@
+const cookie = require('cookie')
+const isNode = require('detect-node')
+
+const createImmutable = require('./helpers/create-immutable')
+
+/**
+ * The main Client class, which provides methods to authenticate against the API and request access tokens.
+ *
+ * @class
+ */
+class Client {
+  /**
+   * Creates a new instance of the client
+   *
+   * @param {object} config - Configuration parameters for the client
+   * @param {object} httpClient - A configured HTTP client for making requests
+   */
+  constructor (config, httpClient) {
+    createImmutable(this, '_config', config)
+    createImmutable(this, '_httpClient', httpClient)
+    this._resetState()
+  }
+
+  /**
+   * Makes a request to exchange a JWT for a access token.
+   *
+   * @async
+   * @private
+   * @param {string} jwt - A JSON Web Token
+   * @returns {Promise} Resolves to the new access token.
+   */
+  async _exchangeJWTForToken (jwt) {
+    if (!jwt) {
+      return null
+    }
+
+    const data = {
+      'client_id': this._config.clientAppID,
+      'grant_type': 'password'
+    }
+
+    const config = {
+      headers: {
+        cookie: `${this._config.cookieName}=${jwt}`
+      },
+      withCredentials: true
+    }
+
+    const response = await this._httpClient.post('/oauth/token', data, config)
+    this._handleNewTokenData(response.data)
+    return this._state.accessToken
+  }
+
+  /**
+   * Gets an access token. Returns the current access token if it's still valid, otherwise attempts to try and refresh it before returning.
+   *
+   * @async
+   * @returns {Promise} Resolves to the access token, or an empty string if not available.
+   */
+  async getAccessToken() {
+    if (!this._state.accessToken) {
+      return ''
+    } else if (this._isAccessTokenValid()) {
+      return this._state.accessToken
+    } else {
+      return this._refreshTokenData()
+    }
+  }
+
+  /**
+   * Requests a cross-site request forgery (CSRF) token, which is used in the main authentication-related actions.
+   *
+   * @async
+   * @private
+   * @returns {Promise} Resolves to the CSRF token.
+   */
+  async _getCSRFToken () {
+    try {
+      const url = `/users/sign_in?now=${Date.now()}`
+      const config = { withCredentials: true }
+      const response = await this._httpClient.head(url, config)
+      const csrfToken = response.headers['x-csrf-token']
+      return csrfToken
+    } catch (error) {
+      console.error(error)
+      throw error
+    }
+  }
+
+  /**
+   * Tries to get the Panoptes session cookie from the browser's cookie jar.
```

Comment:
Panoptes sessions use HttpOnly cookies so I’m not sure how this works.

Reply:
I believe Rog last worked with @camallen to get a development version of this working locally with Panoptes, so I'm mentioning @camallen to help answer your question.

Reply:
When i was working with Rog was getting the browser to send the cookie to the server side node app. We got that working similar to how we solved the CFE single sign on. 

I assume this code was meant to run server side, though the comments suggest browser 🤔 - indeed our cookies are secure and http only client side in the browser.

Reply:
This gets the cookie from `document.cookie` so it must be run in a browser and, yeah, we deliberately hide our session cookies from `document.cookie` so that they aren't visible to XSS attacks.

Reply:
My understanding is this client is supposed to work server side and client side. There may be separate pathways for those use cases, but I'm not sure.

Reply:
Looking over the proposal (noted in the PR description: https://github.com/zooniverse/front-end-monorepo/pull/1361/files) it's not explicitly stated, but I recall that being the stated goal, a working client for server side and client side, so that it can function for the long term goal of working server side auth and be a replacement of the old auth client for the SPAs we have. Happy to revisit this if this goal is too broad because working server side auth is more important for the long term goals of the monorepo, imo.

Reply:
Server-side auth would work by getting the session secret from the incoming request cookies, then trading the session for a token that can be used as an Auth header. NextJS has an auth example for this, which I've found really useful.
https://github.com/vercel/next.js/blob/7c203b80c98d5b2f5a2af41d29f8032b1242e9c9/examples/with-cookie-auth-fauna/pages/api/profile.js#L10-L19

The docs for the `resumeSession` method mention getting the session cookie from the request, but that code isn't in this PR.
https://github.com/zooniverse/front-end-monorepo/blob/a7ebdfa2adb875a7e862290857a2440e804bb148/packages/lib-auth/src/client.js#L253

Client-side auth is still puzzling me, because the implementation here won't work (and shouldn't work because Panoptes sessions should not be accessible from JS.) This client seems, to me, like it would only work if you re-authenticate with your username and password on each request. I feel like I'm missing something obvious.